compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingInterfaces.td

```diff
@@ -0,0 +1,48 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_DIALECT_ENCODING_INTERFACES
+#define IREE_DIALECT_ENCODING_INTERFACES
+
+include "iree/compiler/Dialect/Encoding/IR/EncodingBase.td"
+include "mlir/IR/BuiltinAttributeInterfaces.td"
+
+def IREEEncoding_EncodingLayoutAttrInterface :
+  AttrInterface<"EncodingLayoutAttrInterface"> {
+  let cppNamespace = "::mlir::iree_compiler::IREE::Encoding";
+  let description = [{
+    Interface used to query layout information needed to materialize encoding
+    attributes.
+
+    Any backend can implement the interface to interpret an encoding layout
+    based on their needs.
+
+    TBD. The current expectation of the interface is to propagate layout
+    information from backends to the host compliation or other targets.
+  }];
+
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/[{
+        Returns the storage size (in bytes) for the tensor types with an
+        optional encoding.
+      }],
+      /*retTy=*/"::mlir::OpFoldResult",
+      /*methodName=*/"calculateStorageSizeInBytes",
+      /*args=*/(ins
+        "::mlir::OpBuilder &":$builder,
+        "RankedTensorType":$type,
+        "ValueRange":$dynamicDims
+      ),
+      /*methodBody=*/"",
+      /*defaultImplementation=*/[{
+        return {};
```

Comment:
Would it be better to insert an assert to fail here? I assume it must be re-implemented by whoever inherits the interface.

Reply:
We could do it. The other way to think about it is that the caller can check if the return value is nullptr or not; raise the error with proper messages. Adding an assertion is being careful to me, while it could add confusion to users because the error messages and stack dump are all bad. Both work for me because we do not have a working case today..

Reply:
As a user I would definitely prefer not to check null every time I call this function, if such API call guarantees that it will definitely return something sensible other than nullptr.

Reply:
sg, updated